crates/iceberg-catalog/src/service/endpoint_statistics.rs

```diff
@@ -0,0 +1,291 @@
+                query_params,
+            })
+            .await
+        {
+            tracing::error!("Failed to send endpoint statistics message: {}", e);
+        };
+    } else {
+        tracing::error!(?path_params, "No request metadata found.");
+    }
+
+    response
+}
+
+/// Sender for the endpoint statistics tracker.
+#[derive(Debug, Clone)]
+pub struct EndpointStatisticsTrackerTx(tokio::sync::mpsc::Sender<EndpointStatisticsMessage>);
+
+impl EndpointStatisticsTrackerTx {
+    #[must_use]
+    pub fn new(tx: tokio::sync::mpsc::Sender<EndpointStatisticsMessage>) -> Self {
+        Self(tx)
+    }
+
+    /// Send a message to the endpoint statistics tracker.
+    ///
+    /// # Errors
+    /// If the receiver has been dropped.
+    pub async fn send(
+        &self,
+        msg: EndpointStatisticsMessage,
+    ) -> Result<(), tokio::sync::mpsc::error::SendError<EndpointStatisticsMessage>> {
+        self.0.send(msg).await
+    }
+}
+
+#[derive(Debug)]
+pub enum EndpointStatisticsMessage {
+    EndpointCalled {
+        request_metadata: RequestMetadata,
+        response_status: StatusCode,
+        path_params: HashMap<String, String>,
+        query_params: HashMap<String, String>,
+    },
+    Flush,
+    Shutdown,
+}
+
+#[derive(Debug, Default)]
+pub struct ProjectStatistics {
+    stats: HashMap<EndpointIdentifier, AtomicI64>,
+}
+
+impl ProjectStatistics {
+    #[must_use]
+    pub fn into_consumable(self) -> HashMap<EndpointIdentifier, i64> {
+        self.stats
+            .into_iter()
+            .map(|(k, v)| (k, v.load(std::sync::atomic::Ordering::Relaxed)))
+            .collect()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct EndpointIdentifier {
+    pub uri: Endpoints,
+    pub status_code: StatusCode,
+    pub warehouse: Option<WarehouseIdent>,
+    // probably only relevant for config calls
+    pub warehouse_name: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum FlushMode {
+    Automatic,
+    Manual,
+}
+
+#[derive(Debug)]
+pub struct EndpointStatisticsTracker {
+    rcv: tokio::sync::mpsc::Receiver<EndpointStatisticsMessage>,
+    endpoint_statistics: HashMap<ProjectId, ProjectStatistics>,
+    statistic_sinks: Vec<Arc<dyn EndpointStatisticsSink>>,
+    flush_interval: Duration,
+    flush_mode: FlushMode,
+}
+
+impl EndpointStatisticsTracker {
+    #[must_use]
+    pub fn new(
+        rcv: tokio::sync::mpsc::Receiver<EndpointStatisticsMessage>,
+        stat_sinks: Vec<Arc<dyn EndpointStatisticsSink>>,
+        flush_interval: Duration,
+        flush_mode: FlushMode,
+    ) -> Self {
+        Self {
+            rcv,
+            endpoint_statistics: HashMap::new(),
+            statistic_sinks: stat_sinks,
+            flush_interval,
+            flush_mode,
+        }
+    }
+
+    pub async fn run(mut self) {
+        let mut last_update = tokio::time::Instant::now();
+        loop {
+            if !matches!(self.flush_mode, FlushMode::Manual)
```

Comment:
Should we open a span for all the traces in this function?
This doesn't run as part of an API span, so adding one for this call of the sheduled funciton would be helpful to group all related messages.
We are doing the same in middle for a refresh run.